website/docs/r/app.html.markdown

```diff
@@ -0,0 +1,104 @@
+---
+layout: "rancher2"
+page_title: "Rancher2: rancher2_app"
+sidebar_current: "docs-rancher2-resource-app"
+description: |-
+  Provides a Rancher v2 app resource. This can be used to deploy apps within rancher v2 projects.
+---
+
+# rancher2\_app
+
+Provides a Rancher v2 app resource. This can be used to deploy apps within rancher v2 projects.
+
+This resource can also modify rancher v2 apps in 3 ways:
+- `Update`: If `description`, `annotations` or `labels` arguments are modified the app will be updated. No new `revision_id` will be generated at rancher.
+- `Upgrade`: If `answers`, `external_id` or `values_yaml` arguments are modified, the app will be upgraded. A new `revision_id` will be generated at rancher.
+- `Rollback`: If `revision_id` argument is provided or modified the app will be rollbacked accordingly. A new `revision_id` will be generated at rancher. It will also generate a non empty tf plan that will require manual tf file edition. Use carefully.
+
+Note: In case of multiple resource modification in a row, `rollback` has preference over `upgrade`.
+
+## Example Usage
+
+```hcl
+# Create a new rancher2 App
+resource "rancher2_app" "foo" {
+  name = "foo"
+  description = "Foo app"
+  project_id = "<project_id>"
+  external_id = "catalog://?catalog=<catalog_name>&template=<template_name>&version=<template_version>"
+  target_namespace = "<namespace_name>"
+  answers = {
+    "ingress_host" = "test.xip.io"
+    "foo" = "bar"
+    "ingress.annotations.nginx.ingress.kubernetes.io/force-ssl-redirect" = true
+  }
+}
+```
+
+```hcl
+# Create a new rancher2 App in a new namespace
+resource "rancher2_namespace" "foo" {
+  name = "foo"
+  description = "Foo namespace"
+  project_id = "<project_id>"
+  resource_quota {
+    limit {
+      limits_cpu = "100m"
+      limits_memory = "100Mi"
+      requests_storage = "1Gi"
+    }
+  }
+}
+
+resource "rancher2_app" "foo" {
+  name = "foo"
+  description = "Foo app"
+  project_id = "<project_id>"
+  external_id = "catalog://?catalog=<catalog_name>&template=<template_name>&version=<template_version>"
+  target_namespace = "${rancher2_namespace.foo.name}"
+  answers = {
+    "ingress_host" = "test.xip.io"
+    "foo" = "bar"
+    "ingress.annotations.nginx.ingress.kubernetes.io/force-ssl-redirect" = true
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `external_id` - (Required) The url of the app template on a catalog. If modified, app will be upgraded (string)
```

Comment:
I think we need another solution for this. Expecting the user to craft this weird url is terrible UX. It would be better to have 3 separate args for `catalog_name`, `template_name` (which should be called something more obvious like `app_name` or `chart_name`), and `template_version` (which again should probably be called something more obvious like `chart_version` or `app_version`).

Reply:
I agree, can we construct the id via "catalog_name", "template_name", "version" ?

Reply:
If we're passing version, should we allow for it to default to latest like in the UI?

Reply:
Yeah I think it makes sense to have `version` be optional.

Reply:
Agreed to use `catalog_name`, `template_name` and `template_version` to construct `external_id` and facilitate UX.

For me, makes no sense to let `version` be optional. This is a generic label, not reproducible. Also, how to manage app upgrade?? what would happens when the chart is upgraded on the helm repo?? The app will be set to be updated on next terraform apply.

Reply:
Yeah, that would be the expected behavior I think. That's how the helm provider does it as well: https://www.terraform.io/docs/providers/helm/release.html

If `version` is not specified, it installs the latest every time terraform runs (which does nothing if you are already on the latest version).

Reply:
The problem is not when tf does nothing, the problem is when the helm char is updated and terraform is run, the app will be marked to be upgraded, without any change on tf file.

Also, on rancher API you must specify the version on the `external_id`

Reply:
What do you think if we use a format like multi cluster app, `<catalog_name>-<template_name>-<template_version>`??

Reply:
Yeah, I wouldn't want that to happen, but that is how the helm provider works. I believe the rancher cli also works that way. If you do not specify the version, it will install the latest.

Reply:
I agree it is bad form to not specify a version, but there is enough prior art out there that suggests no version means installing the latest all the time. However, in the case of demo environments and other such non-sense that won't last long, it has some value installing the latest.

I think in this case, TF provides a reasonable method to see that it is going to do something with the plan/apply steps being discrete. At that point, adding a version == to the installed version should mitigate any future changes until desired. 

As far as x-y-z as a string.. I prefer the fields be discrete. If there is enough prior art to say this is the TF way, so be it, but no.

Reply:
Separate fields is definitely more idiomatic.

Reply:
ok guys, separated fields and no version allowed.